client/app/components/ApplicationArea/withUserSession.jsx

```diff
@@ -0,0 +1,67 @@
+import React, { useEffect, useState } from "react";
+import PropTypes from "prop-types";
+import ErrorBoundary, { ErrorBoundaryContext } from "@/components/ErrorBoundary";
+import { Auth } from "@/services/auth";
+import organizationStatus from "@/services/organizationStatus";
+import ApplicationHeader from "./ApplicationHeader";
+import ErrorMessage from "./ErrorMessage";
+
+export default function withUserSession(WrappedComponent) {
+  function UserSessionWrapper({ bodyClass, ...props }) {
+    const [isAuthenticated, setIsAuthenticated] = useState(!!Auth.isAuthenticated());
+
+    useEffect(() => {
+      let isCancelled = false;
+      Promise.all([Auth.requireSession(), organizationStatus.refresh()])
+        .then(() => {
+          if (!isCancelled) {
+            setIsAuthenticated(!!Auth.isAuthenticated());
+          }
+        })
+        .catch(() => {
+          if (!isCancelled) {
+            setIsAuthenticated(false);
+          }
+        });
+      return () => {
+        isCancelled = true;
+      };
+    }, []);
+
+    useEffect(() => {
+      if (bodyClass) {
+        document.body.classList.toggle(bodyClass, true);
+        return () => {
+          document.body.classList.toggle(bodyClass, false);
+        };
+      }
+    }, [bodyClass]);
+
+    if (!isAuthenticated) {
+      return null;
+    }
+
+    return (
+      <>
+        <ApplicationHeader />
+        <ErrorBoundary renderError={error => <ErrorMessage error={error} />}>
+          <ErrorBoundaryContext.Consumer>
+            {({ handleError }) => <WrappedComponent onError={handleError} {...props} />}
```

Comment:
This is definitely better, but I wonder: why a HOC instead of just a regular component that will receive the rendered page as children? Is it to support the ability to pass in the `handleError` function or there is some other reason?

Reply:
I just realized another thing: this way the header will get re-rendered every time, including running its load (load favorites) every time.

Wouldn’t it be better to render a single header and only replace the page itself? I think we can achieve this if we have a single wrapper rendered by the Router and then only replace the page.

Not sure about the downsides though.

Reply:
I'll try to answer for both questions in this comment.

HoC is a compromise between flexibility and amount of code when defining a route. Of course we can make router accept a `component: SomePage` field instead of `render: () => { ... }`, but having a render function allows to more precisely manage how React will update page controller (using `key` props where needed), the way we pass props to page controller (in render function we can spread route params, or pass entire object, add or omit some params, etc.), do conditional rendering if needed, etc. That features are already used, say, for Alert page - where we have a single component for new/view/edit pages and just pass a different `mode` param - to achieve the same with passing just a component and rendering it in Router we have to keep `resolve` thing (that adds new values to route params) - but I want to get rid of it as well.

Another reason is what you mentioned - with HoC we can pass error handler to wrapper components. Also, I'm not sure if it's a good idea to combine it with auth stuff, maybe `withErrorHander` HoC will be better? Anyway, ATM I don't want to add more wrappers, that two auth wrappers are required for all pages, so I think we can keep error handlers there.

Also, HoC is a good choice because we already wrap _most_ of pages with other HoCs, so adding one more wrapper looks not that bad :slightly_smiling_face:

Reply:
> I just realized another thing: this way the header will get re-rendered every time, including running its load (load favorites) every time.

I'll check this, but I think it can be fixed much easier - just by setting `key` prop to header (but I need to try it to be sure :slightly_smiling_face:)

Reply:
I was actually thinking about it returning the render function instead of a HoC, but it ends up with not that much of a difference. How about moving this `key` to it too (most or all pages seem to use it), in case you need it different, it's just matter of making it replaceable.

Reply:
@gabrieldutra That may be a good idea, but I'll explore it tomorrow (for me it's quite late time already)

Reply:
Actually it may be a really cool idea, especially in combination with what @arikfr suggested :thinking: I'll add one more commit tomorrow - will see how it goes